File: 알고리즘 연습/조합순열/Permutation.py
# 라이브러리 사용
from itertools import permutations

result = list(permutations([1,2,3,4],2))
print(result)


# 직접 구현
arr = [1,2,3,4]
result = []
def perm(result, current, arr, k) :
    if k==0 :
        result.append(current.copy()) # 깊은복사로 새로운 배열로 저장해야함
        return

    for i in range(len(arr)) :
        if arr[i] in current :
            continue
        current.append(arr[i])
        perm(result, current, arr, k-1)
        current.pop()
    
perm(result, [], arr, 2)
print(result)

File: 알고리즘 연습/조합순열/Permutation.java
import java.util.ArrayList;
import java.util.List;

public class Permutation {
        public static void main(String[] args) {
        int[] arr = {1,2,3,4};
        List<List<Integer>> result = new ArrayList<>();
        perm(result, new ArrayList<>(), arr, 2);
        System.out.println(result);
    }

    // 배열 arr에서 k개의 값을 가지는 조합 추출
    static void perm(List<List<Integer>> result, List<Integer> current, int[] arr, int k) {
        if (k == 0) {
            result.add(new ArrayList<>(current)); // 리스트안에 리스트
            return;
        }

        for (int i=0; i<arr.length; i++) {
            if (current.contains(arr[i])) { // 중복체크 (이 부분 제거하면 중복가능한 순열)
                continue;
            }
            current.add(arr[i]);
            perm(result, current, arr, k-1);
            current.remove(current.size()-1);
        }
    }

}